lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HadoopPaths.java

```diff
@@ -31,12 +30,22 @@ public static Path hadoopPath(Location location)
         String path = location.toString();
         Path hadoopPath = new Path(path);
         if ("s3".equals(hadoopPath.toUri().getScheme()) && !path.equals(hadoopPath.toString())) {
-            if (hadoopPath.toUri().getFragment() != null) {
-                throw new IllegalArgumentException("Unexpected URI fragment in path: " + path);
-            }
-            URI uri = URI.create(path);
-            return new Path(uri + "#" + URLEncoder.encode(uri.getPath(), UTF_8));
+            return new Path(toPathEncodedUri(location));
         }
         return hadoopPath;
     }
+
+    private static URI toPathEncodedUri(Location location)
+    {
+        try {
+            return new URI(
+                    location.scheme().orElse(null),
+                    location.host().orElse(null),
+                    "/" + location.path(),
+                    location.path());
+        }
+        catch (URISyntaxException e) {
+            throw new VerifyException("Failed to convert location to URI: " + location, e);
```

Comment:
we're not converting location to URI.
we're using URI only as a vehicle to initialize Path so that it carries a fragment

let this method do full conversion, ie move `new Path(URI)` here
and then change the message to "... to Path"